core/api/src/main/java/org/trellisldp/api/BinaryService.java

```diff
@@ -36,6 +39,26 @@
      */
     CompletableFuture<Void> setContent(BinaryMetadata metadata, InputStream stream);
 
+    /**
+     * Set the content for a binary object using a digest algorithm.
+     *
+     * @implSpec The default implementation will compute a digest while processing the {@code InputStream}.
+     * @param metadata the binary metadata
+     * @param stream the context
+     * @param algorithm the digest algorithm
+     * @return the new completion stage containing the server-computed digest.
+     */
+    default CompletableFuture<byte[]> setContent(final BinaryMetadata metadata, final InputStream stream,
+            final String algorithm) {
+        try {
+            final DigestInputStream input = new DigestInputStream(stream, MessageDigest.getInstance(algorithm));
+            return setContent(metadata, input)
+                .thenApply(future -> input.getMessageDigest().digest());
+        } catch (final NoSuchAlgorithmException ex) {
+            throw new IllegalArgumentException("Invalid message digest", ex);
```

Comment:
Actually, what would be even better would be to change the `String algorithm` argument to `MessageDigest algorithm` (also in the `caclulateDigest` method). This would allow the HTTP layer to handle the `NoSuchAlgorithmException` more easily (i.e. entirely in the HTTP layer). With an `IllegalArgumentException` tucked inside a default method, this seems a bit brittle.

Reply:
Oh, nice. My only concern would be that you can (in theory) reset and reuse a `MessageDigest`, which is why I didn't suggest it (but a `byte[]`) instead of a `String`. I suppose that's a security concern, at some level, but…

Reply:
Just to clarify, when I suggested `MessageDigest`, I meant as an argument, not a return value. Thus, the signature would be:

    CompletableFuture<byte[]> calculateDigest(IRI identifier, MessageDigest digest);

and:

    CompletableFuture<byte[]> setContent(BinaryMetadata metadata, InputStream stream,
            MessageDigest digest)

Reply:
Oh, gotcha, that makes sense, and I like it! The exception handling makes much more sense in the HTTP layer.